lib/message_queue.cpp

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2020 - 2022 Pionix GmbH and Contributors to EVerest
+#include <thread>
+
+#include <fmt/format.h>
+
+#include <everest/logging.hpp>
+
+#include <utils/message_queue.hpp>
+
+namespace Everest {
+
+Message::Message(const std::string& topic, const std::string& payload) : topic(topic), payload(payload) {
+}
+
+MessageQueue::MessageQueue(const std::function<void(std::shared_ptr<Message> message)>& message_callback) :
+    message_callback(message_callback), running(true) {
+    this->worker_thread = std::thread([this]() {
+        while (this->running) {
+            std::shared_ptr<Message> message;
+            std::unique_lock<std::mutex> lock(this->message_mutex);
+            this->cv.wait(lock, [this]() { return !this->message_queue.empty(); });
+
+            message = this->message_queue.front();
+            this->message_queue.pop();
+            lock.unlock();
+
+            // pass the message to the message callback
+            this->message_callback(message);
+        }
+    });
+}
+
+void MessageQueue::add(std::shared_ptr<Message> message) {
+    {
+        std::lock_guard<std::mutex> lock(this->message_mutex);
+        this->message_queue.push(message);
+    }
+    this->cv.notify_all();
+}
+
+void MessageQueue::stop() {
+    this->running = false;
+}
+
+MessageHandler::MessageHandler() : running(true) {
+    this->handler_thread = std::thread([this]() {
+        while (this->running) {
+            std::shared_ptr<json> message;
+            std::unique_lock<std::mutex> lock(this->message_mutex);
+            this->cv.wait(lock, [this]() { return !this->message_queue.empty(); });
+
+            message = this->message_queue.front();
+            this->message_queue.pop();
+            lock.unlock();
+
+            auto data = *message.get();
+
+            // get the registered handlers
+            std::vector<std::shared_ptr<TypedHandler>> local_handlers;
```

Comment:
performance wise (although it doesn't matter here), it would be better to copy construct the vector directly instead of iterating and incrementing it one by one